### PR TITLE
Refs #11361 - Allows to set retry count a timeout interval on a job

### PR DESCRIPTION
**Pull request by adamruzicka**
timeouts require theforeman/foreman-tasks#132

### CODE REVIEW

File: app/lib/actions/remote_execution/run_proxy_command.rb [hunk 1]
```diff
@@ -23,12 +66,62 @@ def rescue_strategy
...
+        end
+      end
+
+      def format_exception(exception)
+        { output[:proxy_task_id] =>
+          { :exception_class => exception.class.name,
+            :execption_message => exception.message } }
+      end
+
+      def with_connection_error_handling(event = nil)
```
Comment (iNecas) [inline app/lib/actions/remote_execution/run_proxy_command.rb]: Shouldn't this be part of foreman-tasks, as it's more about the connection between forman-tasks and smart-proxy-dynflow, than about the remote execution itself?

File: app/lib/actions/remote_execution/run_proxy_command.rb [hunk 2]
```diff
@@ -5,17 +5,60 @@ class RunProxyCommand < Actions::ProxyAction
...
+                              time,
+                              Timeout.new
+            metadata[:timeout] = time.to_s
+          end
+          super
+        end
+      end
+
+      def check_task_status
+        if output[:proxy_task_id]
```
Comment (iNecas) [inline app/lib/actions/remote_execution/run_proxy_command.rb]: This logic could also go into https://github.com/theforeman/foreman-tasks/pull/132/files#diff-03dde90ec7e9caa322015dee15f4e956R51

File: app/lib/actions/remote_execution/run_proxy_command.rb [hunk 3]
```diff
@@ -5,17 +5,60 @@ class RunProxyCommand < Actions::ProxyAction
...
+
+      def trigger_proxy_task
+        suspend do |suspended_action|
+          output[:metadata] ||= {}
+          unless metadata[:timeout]
+            time = Time.now + input[:connection_options][:timeout]
+            @world.clock.ping suspended_action,
+                              time,
+                              Timeout.new
+            metadata[:timeout] = time.to_s
```
Comment (iNecas) [inline app/lib/actions/remote_execution/run_proxy_command.rb]: Another candidate for getting into foreman-tasks, and just providing the timeout data via some method

**Comment by iNecas** [comment]
Would it be possible to have at least configuration option to set the default parameters?

**Comment by iNecas** [comment]
How it should behave, when the host is not available at the moment (but will be in a minute?)

**Comment by iNecas** [comment]
What should happen, when the task starts on the proxy and I restart the proxy then?

**Comment by adamruzicka** [comment]
Having that possibility would be nice, I'll look into it.

With the default settings it would try running the command of the host 4 times in 15 second intervals and fail afterwards. The user can then see that it failed with 'host unreachable', bring the host up and rerun the job on the failed hosts. The user could also tweak the connection options for the job (reminds me there's no UI for that) to be tolerant to minor downtimes.

The way it is now it would wait for the timeout period for the proxy task to finish, after which it would ask the proxy about the tasks' status, get 404 as response and fail. This could be fixed by polling the proxy about the status more frequently or making the proxy report to foreman that it is back online and restarting all the tasks that should be running on that proxy.

**Comment by iNecas** [comment]
+1 for making the intervals configurable (global configuration so far).

I'm ok with not retrying on host not being available.

I don't think having the same interval for script startup and script finishing is a good idea, I would suggest for now just applying the timeout on just start of it.

**Comment by ares** [comment]
Please rebase the branch.

**Comment by adamruzicka** [comment]
The issues have been addressed in theforeman/foreman-tasks#139

**Comment by iNecas** [comment]
Tested and works as expected. Please squash and update dependency on foreman-tasks to 0.7.5, as I released new version with the https://github.com/theforeman/foreman-tasks/pull/139

**Comment by iNecas** [comment]
Thanks @adamruzicka

**Pull request merged**